File: backend/app/schemas/form_schema.py
"""Pydantic models for the Supform **Form Schema**.

This is the Python mirror of ``packages/form-schema/schema/form.schema.json`` and is the
authoritative in-code representation of a form. The form engine validates definitions and
submissions against these models. Keep this file and the JSON Schema in lock-step.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

# An i18n string is either a plain string or a {locale: text} map.
I18nString = str | dict[str, str]


class ElementType(StrEnum):
    """Core element types. The set is intentionally open — the engine tolerates unknown
    types (treated as generic value fields) so the platform stays flexible."""

    # text-ish
    TEXT = "text"
    LONGTEXT = "longtext"
    EMAIL = "email"
    URL = "url"
    PHONE = "phone"
    # numeric
    NUMBER = "number"
    INTEGER = "integer"
    DECIMAL = "decimal"
    # choice
    SINGLE_CHOICE = "single_choice"
    MULTI_CHOICE = "multi_choice"
    DROPDOWN = "dropdown"
    RANKING = "ranking"
    RATING = "rating"
    SCALE = "scale"
    # date/time/bool
    DATE = "date"
    TIME = "time"
    DATETIME = "datetime"
    DATE_RANGE = "date_range"
    BOOLEAN = "boolean"
    # complex
    MATRIX = "matrix"
    GROUP = "group"
    REPEAT = "repeat"
    # media
    FILE = "file"
    IMAGE = "image"
    SIGNATURE = "signature"
    ADDRESS = "address"
    GEOPOINT = "geopoint"
    GEOTRACE = "geotrace"
    GEOSHAPE = "geoshape"
    BARCODE = "barcode"
    # metadata auto-capture (invisible to respondents, filled server-side)
    START = "start"
    END = "end"
    TODAY = "today"
    DEVICEID = "deviceid"
    USERNAME = "username"
    # derived / layout
    CALCULATED = "calculated"
    HIDDEN = "hidden"
    NOTE = "note"
    SECTION = "section"
    HTML = "html"


class Choice(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    value: str | int | float | bool
    label: I18nString | None = None
    visible_if: str | None = Field(default=None, alias="visibleIf")
    score: float | None = None  # points awarded when chosen (quiz mode)
    correct: bool | None = None  # marks a correct answer (quiz mode grading)
    meta: dict[str, Any] = Field(default_factory=dict)


class Feedback(BaseModel):
    """Quiz mode: messages shown on the results screen after grading a question."""

    model_config = ConfigDict(extra="forbid")

    correct: I18nString | None = None
    incorrect: I18nString | None = None


class Outcome(BaseModel):
    """A scored-result band: shown on the thank-you screen when score is in [min, max]."""

    model_config = ConfigDict(populate_by_name=True)

    min: float
    max: float
    message: I18nString
    redirect_url: str | None = Field(default=None, alias="redirectUrl")


class Validation(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    min: float | None = None
    max: float | None = None
    min_length: int | None = Field(default=None, alias="minLength")
    max_length: int | None = Field(default=None, alias="maxLength")
    pattern: str | None = None
    min_selected: int | None = Field(default=None, alias="minSelected")
    max_selected: int | None = Field(default=None, alias="maxSelected")
    expression: str | None = None
    message: I18nString | None = None


class RepeatSettings(BaseModel):
    min: int = 0
    max: int | None = None
    add_button_text: I18nString | None = Field(default=None, alias="addButtonText")
    entry_label: I18nString | None = Field(default=None, alias="entryLabel")


class Element(BaseModel):
    """A field, layout block, group, or repeat. ``type`` drives behavior."""

    model_config = ConfigDict(populate_by_name=True, extra="allow")

    type: str
    name: str
    label: I18nString | None = None
    hint: I18nString | None = None
    placeholder: I18nString | None = None
    default_value: Any | None = Field(default=None, alias="defaultValue")
    required: bool = False
    read_only: bool = Field(default=False, alias="readOnly")

    # Logic (expressions evaluated by app.form_engine.expressions)
    visible_if: str | None = Field(default=None, alias="visibleIf")
    enable_if: str | None = Field(default=None, alias="enableIf")
    required_if: str | None = Field(default=None, alias="requiredIf")
    calculate: str | None = None

    validation: Validation | None = None

    # choices / matrix
    options: list[Choice] | None = None
    options_from: str | None = Field(default=None, alias="optionsFrom")
    rows: list[Choice] | None = None
    columns: list[Choice] | None = None

    # nesting
    elements: list[Element] | None = None
    repeat: RepeatSettings | None = None

    # quiz grading (see app.form_engine.scoring)
    points: float | None = None  # points for a correct answer (default 1 when graded)
    correct_answer: bool | int | float | str | list[Any] | None = Field(
        default=None, alias="correctAnswer"
    )
    feedback: Feedback | None = None

    meta: dict[str, Any] = Field(default_factory=dict)


class NextPageRule(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    condition: str
    page: str


class Page(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    name: str
    title: I18nString | None = None
    description: I18nString | None = None
    visible_if: str | None = Field(default=None, alias="visibleIf")
    next_page_if: list[NextPageRule] = Field(default_factory=list, alias="nextPageIf")
    elements: list[Element] = Field(default_factory=list)


class Theme(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="allow")

    preset: str = "supform-light"
    primary_color: str | None = Field(default=None, alias="primaryColor")
    background_color: str | None = Field(default=None, alias="backgroundColor")
    font_family: str | None = Field(default=None, alias="fontFamily")
    corner_radius: int | None = Field(default=None, alias="cornerRadius")
    cover_image: str | None = Field(default=None, alias="coverImage")
    logo: str | None = None


class QualityChecks(BaseModel):
    """Thresholds for automated data-quality flags computed at submission time."""

    model_config = ConfigDict(populate_by_name=True, extra="allow")

    min_duration_seconds: int | None = Field(default=None, alias="minDurationSeconds")
    expected_geo_bbox: list[float] | None = Field(default=None, alias="expectedGeoBbox")


class FormSettings(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="allow")

    display_mode: str = Field(default="paged", alias="displayMode")
    show_progress_bar: bool = Field(default=True, alias="showProgressBar")
    shuffle_questions: bool = Field(default=False, alias="shuffleQuestions")
    shuffle_options: bool = Field(default=False, alias="shuffleOptions")
    allow_multiple_submissions: bool = Field(default=True, alias="allowMultipleSubmissions")
    require_login: bool = Field(default=False, alias="requireLogin")
    accepting_responses: bool = Field(default=True, alias="acceptingResponses")
    open_date: str | None = Field(default=None, alias="openDate")
    close_date: str | None = Field(default=None, alias="closeDate")
    max_responses: int | None = Field(default=None, alias="maxResponses")
    submit_button_text: I18nString | None = Field(default=None, alias="submitButtonText")
    confirmation_title: I18nString | None = Field(default=None, alias="confirmationTitle")
    confirmation_message: I18nString | None = Field(default=None, alias="confirmationMessage")
    welcome_title: I18nString | None = Field(default=None, alias="welcomeTitle")
    welcome_message: I18nString | None = Field(default=None, alias="welcomeMessage")
    redirect_url: str | None = Field(default=None, alias="redirectUrl")
    notify_emails: list[str] = Field(default_factory=list, alias="notifyEmails")
    quiz_mode: bool = Field(default=False, alias="quizMode")
    show_correct_answers: bool = Field(default=True, alias="showCorrectAnswers")
    workflow_steps: list[str] = Field(default_factory=list, alias="workflowSteps")
    outcomes: list[Outcome] = Field(default_factory=list)
    # Data quality: thresholds for automated flag checks run at submit time.
    quality_checks: QualityChecks | None = Field(default=None, alias="qualityChecks")


class FormSchema(BaseModel):
    """The full form definition."""

    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    schema_version: str = Field(default="1.0", alias="schemaVersion")
    id: str | None = None
    name: str
    title: I18nString
    description: I18nString | None = None
    version: int = 1
    default_language: str = Field(default="en", alias="defaultLanguage")
    languages: list[str] = Field(default_factory=list)
    theme: Theme = Field(default_factory=Theme)
    settings: FormSettings = Field(default_factory=FormSettings)
    pages: list[Page] = Field(default_factory=list)

    # ---- convenience ----
    def iter_elements(self) -> list[Element]:
        """Flatten all elements (recursing into groups/repeats)."""
        out: list[Element] = []

        def _walk(elements: list[Element]) -> None:
            for el in elements:
                out.append(el)
                if el.elements:
                    _walk(el.elements)

        for page in self.pages:
            _walk(page.elements)
        return out

    def field_names(self) -> set[str]:
        return {el.name for el in self.iter_elements()}


Element.model_rebuild()

File: frontend/src/types/form-schema.ts
/**
 * TypeScript mirror of packages/form-schema/schema/form.schema.json.
 *
 * This is the frontend half of the form contract. The builder produces these objects;
 * the renderer consumes them. Keep in lock-step with the JSON Schema and the backend
 * Pydantic models (app/schemas/form_schema.py).
 */

/** Either a plain string or a { locale: text } map. */
export type I18nString = string | Record<string, string>;

/** A logic/value expression evaluated by the form engine (e.g. "age >= 18"). */
export type Expression = string;

export type ElementType =
  | "text"
  | "longtext"
  | "email"
  | "url"
  | "phone"
  | "number"
  | "integer"
  | "decimal"
  | "single_choice"
  | "multi_choice"
  | "dropdown"
  | "ranking"
  | "rating"
  | "scale"
  | "date"
  | "time"
  | "datetime"
  | "date_range"
  | "boolean"
  | "matrix"
  | "group"
  | "repeat"
  | "file"
  | "image"
  | "signature"
  | "address"
  | "geopoint"
  | "geotrace"
  | "geoshape"
  | "barcode"
  | "start"
  | "end"
  | "today"
  | "deviceid"
  | "username"
  | "calculated"
  | "hidden"
  | "note"
  | "section"
  | "html"
  // open set: custom types are allowed
  | (string & {});

export interface Choice {
  value: string | number | boolean;
  label?: I18nString;
  visibleIf?: Expression;
  /** Points awarded when this option is chosen (quiz mode). */
  score?: number;
  meta?: Record<string, unknown>;
}

/** A scored-result band shown on the thank-you screen when the score falls in [min, max]. */
export interface Outcome {
  min: number;
  max: number;
  message: I18nString;
  redirectUrl?: string;
}

export interface Validation {
  min?: number;
  max?: number;
  minLength?: number;
  maxLength?: number;
  pattern?: string;
  minSelected?: number;
  maxSelected?: number;
  expression?: Expression;
  message?: I18nString;
}

export interface RepeatSettings {
  min?: number;
  max?: number;
  addButtonText?: I18nString;
  entryLabel?: I18nString;
}

export interface Element {
  type: ElementType;
  name: string;
  label?: I18nString;
  hint?: I18nString;
  placeholder?: I18nString;
  defaultValue?: unknown;
  required?: boolean;
  readOnly?: boolean;
  visibleIf?: Expression;
  enableIf?: Expression;
  requiredIf?: Expression;
  calculate?: Expression;
  validation?: Validation;
  options?: Choice[];
  optionsFrom?: string;
  rows?: Choice[];
  columns?: Choice[];
  elements?: Element[];
  repeat?: RepeatSettings;
  meta?: Record<string, unknown>;
}

export interface NextPageRule {
  condition: Expression;
  page: string;
}

export interface Page {
  name: string;
  title?: I18nString;
  description?: I18nString;
  visibleIf?: Expression;
  /** Conditional branching: first matching rule wins; falls back to sequential if none match. */
  nextPageIf?: NextPageRule[];
  elements: Element[];
}

export interface Theme {
  preset?: string;
  primaryColor?: string;
  backgroundColor?: string;
  fontFamily?: string;
  cornerRadius?: number;
  coverImage?: string;
  logo?: string;
  [key: string]: unknown;
}

export interface FormSettings {
  displayMode?: "paged" | "single" | "oneQuestionPerScreen";
  showProgressBar?: boolean;
  shuffleQuestions?: boolean;
  allowMultipleSubmissions?: boolean;
  requireLogin?: boolean;
  closeDate?: string;
  maxResponses?: number;
  submitButtonText?: I18nString;
  /** Heading shown on the thank-you screen (defaults to "Thank you!"). */
  confirmationTitle?: I18nString;
  confirmationMessage?: I18nString;
  /** Optional welcome screen shown before the first step (paged / one-question modes). */
  welcomeTitle?: I18nString;
  welcomeMessage?: I18nString;
  /** Where to send the respondent after submitting (auto-redirect from the thank-you screen). */
  redirectUrl?: string;
  /** Email addresses notified on each new submission. */
  notifyEmails?: string[];
  /** Score selected-option points and show an outcome on the thank-you screen. */
  quizMode?: boolean;
  outcomes?: Outcome[];
  qualityChecks?: QualityChecks;
  workflowSteps?: string[];
}

export interface QualityChecks {
  /** Submissions completed faster than this many seconds are flagged "too_fast". Default 30. */
  minDurationSeconds?: number;
  /** Geopoint answers outside [minLat, minLng, maxLat, maxLng] are flagged "geo_outlier". */
  expectedGeoBbox?: [number, number, number, number];
}

export interface FormSchema {
  schemaVersion: string;
  id?: string;
  name: string;
  title: I18nString;
  description?: I18nString;
  version?: number;
  defaultLanguage?: string;
  languages?: string[];
  theme?: Theme;
  settings?: FormSettings;
  pages: Page[];
}
